### DISCUSSION TITLE
Possible AsynchScheduler performance issue

### DISCUSSION THREAD

**Discussion by akaltar** [ReactiveX/rxjs#7236] [Report issues other than bug]
Hello! I'm using `rxjs` in a big project through `redux-observable`, and in certain conditions we spend over ~5s in pure `rxjs` code without executing any of our own code.

![perf trace of spending time in rxjs](https://user-images.githubusercontent.com/2000269/227999902-4cc6eb9e-4612-4abc-a37f-84b1375873f5.png)

I believe I might have found the root cause for spending so much time "spinning wheels" in rxjs, and I might be able to come up with an optimization that could make things better, but I want to make sure this aligns with the way development is going.

First, here is a repo with a reproduction of the performance issue: https://github.com/akaltar/rxjs-perf-repro

I think the issue might be in the way the looping is done over the actions in [`AsyncScheduler.ts:flush`](https://github.com/ReactiveX/rxjs/blob/667e185435fbd6d83e92942779501bf960249948/src/internal/scheduler/AsyncScheduler.ts)
+ the way an action removes itself from the parent schedulers.

I'm not very familiar with how rxjs works, but I think replacing the loop with a simple for-loop, and adding a fast-path to removing the action from the parent `actions` array could speed things up a lot wihtout breaking things.

Maybe by not trying to remove `this` from the `actions` array at all, since at that point in `flush` it was already removed by `shift`, in the for-loop scenario, we could maybe just clean up the actions by assigning a fresh empty array after executing.  Should be a non-breaking change and gain quite a bit of performance
I think further gains could be had if actions were stored in a `Map`, but that would be a much bigger change. (A bit like #7234 just for AsyncScheduler)

In either case I'm happy to put together a PR if you don't see any fundamental issues with these suggestions

**Comment by gyhyfj** [comment]
I encountered with another performance issue when I was coding a script for crawling sth. 
When running that script, it didn't cost much memo but always froze later without any error.
Then I replace `from` + `concatMap` API with normal `for`  loop, it worked well again and much faster than before ( down from 20s to 5s per loop)
